pkg/ingester/stream.go

```diff
@@ -106,13 +116,18 @@ func (s *stream) Push(_ context.Context, entries []logproto.Entry) error {
 	for i := range entries {
 		chunk := &s.chunks[len(s.chunks)-1]
 		if chunk.closed || !chunk.chunk.SpaceFor(&entries[i]) {
+			err := chunk.chunk.Close()
```

Comment:
I'm not clear why we need to cut the head block now but we didn't have to before, could you explain?

Reply:
by the nature of how we calculated the SpaceFor previously, it was always guaranteed to have cut the head chunk:
  * append a log line, if that log line is > than the block size, cut the block.
  * next log line, check to see if the number of blocks is >= max blocks per chunk, if yes, make a new chunk.

When we are checking SpaceFor against a target size there head block could still have entries in it but the new line could push us over the edge, so we have to make sure to cut what's out of the head block.

I debated if the SpaceFor could incorporate this behavior but that felt like kind of an ugly side-effect of a method which I wouldn't expect to make any changes to the chunk.

Reply:
good for me.

Reply:
:+1: